helpers/form/ElementMapFactory.php

```diff
@@ -73,45 +140,35 @@ public function create(core_kernel_classes_Property $property): ?tao_helpers_for
             return null;
         }
 
-        $widgetUri   = $widgetResource->getUri();
-        $propertyUri = $property->getUri();
+        $widgetUri      = $widgetResource->getUri();
+        $propertyUri    = $property->getUri();
+        $expectedWidget = $widgetUri;
+
         // Authoring widget is not used in standalone mode
-        if (
-            $widgetUri === Authoring::WIDGET_ID
-            && tao_helpers_Context::check('STANDALONE_MODE')
-        ) {
+        if ($widgetUri === Authoring::WIDGET_ID && $this->isStandaloneMode()) {
             return null;
         }
 
         // Horrible hack to fix file widget
         if ($widgetUri === AsyncFile::WIDGET_ID) {
-            $widgetResource = new core_kernel_classes_Resource(GenerisAsyncFile::WIDGET_ID);
+            $expectedWidget = GenerisAsyncFile::WIDGET_ID;
+            $widgetResource = new core_kernel_classes_Resource($expectedWidget);
         }
 
-        $element = tao_helpers_form_FormFactory::getElementByWidget(
-            tao_helpers_Uri::encode($propertyUri),
-            $widgetResource
-        );
+        if ($this->hasElement) {
+            $element = $this->element;
+        } else {
+            $element = tao_helpers_form_FormFactory::getElementByWidget(
+                tao_helpers_Uri::encode($propertyUri),
+                $widgetResource
+            );
+        }
```

Comment:
In this case, you don't need `$this->hasElement` property anymore.
```suggestion
        $element = $this->element ?? tao_helpers_form_FormFactory::getElementByWidget(
            tao_helpers_Uri::encode($propertyUri),
            $widgetResource
        );
```

Reply:
See https://github.com/oat-sa/tao-core/pull/3384/files#r802472072, this is used to explicitly set the element to `null` in tests while also skipping the call to `getElementByWidget()`.